src/events/etcd.py

```diff
@@ -102,16 +123,56 @@ def _on_cluster_relation_joined(self, event: RelationJoinedEvent) -> None:
         # Todo: remove this test at some point, this is just for showcasing that it works :)
         # We will need to perform any HA-related action against the raft leader
         # e.g. add members, trigger leader election, log compaction, etc.
-        if raft_leader := self.charm.cluster_manager.get_leader():
+        try:
+            raft_leader = self.charm.cluster_manager.get_leader()
             logger.info(f"Raft leader: {raft_leader}")
+        except RaftLeaderNotFoundError as e:
+            logger.warning(e)
 
     def _on_leader_elected(self, event: LeaderElectedEvent) -> None:
         """Handle all events in the 'cluster' peer relation."""
         if not self.charm.state.peer_relation:
             self.charm.set_status(Status.NO_PEER_RELATION)
             return
 
+        if self.charm.unit.is_leader() and not self.charm.state.cluster.internal_user_credentials:
+            self.charm.state.cluster.update(
+                {f"{INTERNAL_USER}-password": self.charm.workload.generate_password()}
+            )
+
     def _on_update_status(self, event: ops.UpdateStatusEvent) -> None:
         """Handle update_status event."""
         if not self.charm.workload.alive():
             self.charm.set_status(Status.SERVICE_NOT_RUNNING)
+
+    def _on_secret_changed(self, event: ops.SecretChangedEvent) -> None:
+        """Handle the secret_changed event."""
+        if not self.charm.unit.is_leader():
+            return
+
+        if admin_secret_id := self.charm.config.get(INTERNAL_USER_PASSWORD_CONFIG):
+            if admin_secret_id == event.secret.id:
+                self.update_admin_password(admin_secret_id)
+
+    def update_admin_password(self, admin_secret_id: str) -> None:
+        """Compare current admin password and update in etcd if required."""
+        try:
+            if new_password := get_secret_from_id(self.charm.model, admin_secret_id).get(
+                INTERNAL_USER_PASSWORD_CONFIG
+            ):
+                # only update admin credentials if the password has changed
+                if new_password != self.charm.state.cluster.internal_user_credentials.get(
+                    INTERNAL_USER
+                ):
+                    logger.debug(f"{INTERNAL_USER_PASSWORD_CONFIG} has changed.")
+                    try:
+                        self.charm.cluster_manager.update_credentials(
+                            username=INTERNAL_USER, password=new_password
+                        )
+                        self.charm.state.cluster.update(
```

Comment:
I believe we need to, for two reasons:
- otherwise we cannot compare if the password has actually changed when we get a `secret_changed` event
- if the secret and config is removed, we would not know the password anymore

Reply:
Yes, I believe this is a design decision. The alternative can be:

- Store the URI on the data bag.
- On `secret_changed` if the the id is the same we update the password always.
- If the secret is removed then we regenerate the password.

I do not mind both approaches but I think the one implemented is better in the case of the `tls-private-key` because always assuming the content changed can add an overhead of re-issueing certificates.

NB: I'm assuming we want to handle all juju user secrets the same way.

Reply:
I also think it is a question of resilience: If the secret is - for whatever reason - not granted to our application anymore, we would not be able to retrieve the password. This would break our operator immediately, because not having the password means not being able to generate a new one.